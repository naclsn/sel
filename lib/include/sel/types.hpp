#ifndef SEL_TYPE_HPP
#define SEL_TYPE_HPP

/**
 * Parsing and representation of types, as well as a set
 * of helper functions to deal with the type system.
 */

#include <ostream>
#include <istream>
#include <string>
#include <vector>
#include <iterator>


namespace sel {

  enum class Ty {
    UNK,
    NUM,
    STR,
    LST,
    FUN,
  };
  enum TyFlag {
    IS_FIN = 0, // (0, and as such default)
    IS_INF = 1,
    IS_TPL = 2,
  };

  /**
   * Represents a type (yey). Use the [..]Type function
   * to construct/parse.
   */
  struct Type {
    Ty base = Ty::UNK;
    union P {
      std::string* name;
      std::vector<Type*>* box_has; // Ty has;
      Type* box_pair[2]; // Ty pair[2];
    } p = {.name=nullptr};
    uint8_t flags = 0;

    Type() { }
    Type(Ty base, Type::P p, uint8_t flags);
    Type(Type const& ty);
    Type(Type&& ty) noexcept;
    ~Type();

    bool operator==(Type const& other) const;
    bool operator!=(Type const& other) const;

    std::vector<Type*>& has() const { return *p.box_has; }
    Type const& from() const { return *p.box_pair[0]; }
    Type const& to() const { return *p.box_pair[1]; }
    Type applied(Type const& arg) const;
  };

  /**
   * Parse a type from the given stream. The overload to
   * the `>>` operator is also provided as a convenience,
   * but this function allows retreiving a name
   * (which would be specified with the `<name> ::`
   * syntax). If `named` is not nullptr but the parsed
   * type representation does not specify a name, the
   * empty string is written into `named`.
  */
  void parseType(std::istream& in, std::string* named, Type& res);

  std::ostream& operator<<(std::ostream& out, Type const& ty);
  std::istream& operator>>(std::istream& in, Type& res);

} // namespace sel

#endif // SEL_TYPE_HPP
