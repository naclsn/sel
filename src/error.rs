use std::fmt::{Display, Formatter, Result as FmtResult};

use crate::parse::{Applicable, TokenKind};
use crate::scope::{Location, SourceRegistry};
use crate::types::FrozenType;

// error types {{{
#[derive(PartialEq, Debug)]
pub enum ErrorContext {
    Unmatched {
        open_token: TokenKind,
    },
    CompleteType {
        complete_type: FrozenType,
    },
    TypeListInferredItemType {
        list_item_type: FrozenType,
    },
    AsNthArgToNowTyped {
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromAsNthArgToNowTyped {
        comma_loc: Location,
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromToNotFunc {
        comma_loc: Location,
    },
    AutoCoercedVia {
        func_name: String,
        func_type: FrozenType,
    },
    DeclaredHereWithType {
        with_type: FrozenType,
    },
    LetFallbackTypeMismatch {
        result_type: FrozenType,
        fallback_type: FrozenType,
    },
}

#[derive(PartialEq, Debug)]
pub enum ErrorKind {
    ContextCaused {
        error: Box<Error>,
        because: ErrorContext,
    },
    Unexpected {
        token: TokenKind,
        expected: &'static str,
    },
    UnexpectedDefInScript,
    UnknownName {
        name: String,
        expected_type: FrozenType,
    },
    NotFunc {
        actual_type: FrozenType,
    },
    TooManyArgs {
        nth_arg: usize,
        func: Applicable,
    },
    ExpectedButGot {
        expected: FrozenType,
        actual: FrozenType,
    },
    InfWhereFinExpected,
    NameAlreadyDeclared {
        name: String,
    },
    CouldNotReadFile {
        error: String,
    },
}

#[derive(PartialEq, Debug)]
pub struct Error(pub Location, pub ErrorKind);

#[derive(Default, PartialEq, Debug)]
pub struct ErrorList(pub(crate) Vec<Error>);

impl IntoIterator for ErrorList {
    type Item = Error;
    type IntoIter = <Vec<Error> as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl ErrorList {
    pub(crate) fn push(&mut self, err: Error) {
        self.0.push(err);
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}
// }}}

pub struct Report<'a> {
    registry: &'a SourceRegistry,
    title: String,
    messages: Vec<(Location, String)>,
}

// generate report {{{
impl Error {
    fn ctx_messages(loc: Location, because: &ErrorContext, report: &mut Report) {
        use ErrorContext::*;
        use TokenKind::*;

        let msgs: &[_] = match because {
            Unmatched { open_token } => &[(
                loc,
                format!(
                    "open {} here",
                    match open_token {
                        OpenBracket => '[',
                        OpenBrace => '{',
                        Unknown(_) | Word(_) | Bytes(_) | Number(_) | Comma | CloseBracket
                        | CloseBrace | Equal | Def | Let | Use | Semicolon | End => unreachable!(),
                    }
                ),
            )],
            CompleteType { complete_type } => &[(loc, format!("complete type: {complete_type}"))],
            TypeListInferredItemType { list_item_type } => &[(
                loc,
                format!("list type was inferred to be [{list_item_type}]"),
            )],
            AsNthArgToNowTyped {
                nth_arg,
                func,
                type_with_curr_args,
            } => &[(
                loc,
                format!(
                    "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                    match nth_arg {
                        1 => "st",
                        2 => "nd",
                        3 => "rd",
                        _ => "th",
                    },
                    match func {
                        Applicable::Name(name) => name,
                        Applicable::Bind(_, _, _) => "let binding",
                    },
                ),
            )],
            ChainedFromAsNthArgToNowTyped {
                comma_loc,
                nth_arg,
                func,
                type_with_curr_args,
            } => &[
                (
                    loc,
                    format!(
                        "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                        match nth_arg {
                            1 => "st",
                            2 => "nd",
                            3 => "rd",
                            _ => "th",
                        },
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                    ),
                ),
                (comma_loc.clone(), "chained through here".into()),
            ],
            ChainedFromToNotFunc { comma_loc } => &[
                (loc, "Not a function".into()),
                (comma_loc.clone(), "chained through here".into()),
            ],
            AutoCoercedVia {
                func_name,
                func_type,
            } => &[(
                loc,
                format!("coerced via '{func_name}' (which has type {func_type})"),
            )],
            DeclaredHereWithType { with_type } => &[(loc, format!("here with type {with_type}"))],
            LetFallbackTypeMismatch {
                result_type,
                fallback_type,
            } => &[(
                loc,
                format!("fallback of type {fallback_type} doesn't match result type {result_type}"),
            )],
        };
        report.messages.extend_from_slice(msgs);
    }

    pub fn report<'a>(&self, registry: &'a SourceRegistry) -> Report<'a> {
        use ErrorKind::*;
        use TokenKind::*;

        let loc = self.0.clone();

        match &self.1 {
            ContextCaused { error, because } => {
                let mut r = error.report(registry);
                Error::ctx_messages(loc, because, &mut r);
                r
            }
            Unexpected { token, expected } => Report {
                registry,
                title: "Unexpected token".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Unexpected '{}', expected {expected}",
                        match token {
                            Unknown(t) => t,
                            Word(v) => v,
                            Bytes(_) => "(bytes)",
                            Number(_) => "(number)",
                            Comma => ",",
                            OpenBracket => "[",
                            CloseBracket => "]",
                            OpenBrace => "{",
                            CloseBrace => "}",
                            Equal => "=",
                            Def => "def",
                            Let => "let",
                            Use => "use",
                            Semicolon => ";",
                            End => "end of script",
                        }
                    ),
                )],
            },
            UnexpectedDefInScript => Report {
                registry,
                title: "Unexpected definition".into(),
                messages: vec![(
                    loc,
                    "these should only appear before the script, at the beginning of the file"
                        .into(),
                )],
            },
            UnknownName {
                name,
                expected_type,
            } => Report {
                registry,
                title: "Unknown name".into(),
                messages: vec![(
                    loc,
                    format!("Unknown name '{name}', should be of type {expected_type}"),
                )],
            },
            NotFunc { actual_type } => Report {
                registry,
                title: "Not a function".into(),
                messages: vec![(
                    loc,
                    format!("Expected a function type, but got {actual_type}"),
                )],
            },
            TooManyArgs { nth_arg, func } => Report {
                registry,
                title: "Too many argument".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Too many argument to {}, expected only {}",
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                        nth_arg - 1
                    ),
                )],
            },
            ExpectedButGot { expected, actual } => Report {
                registry,
                title: "Type mismatch".into(),
                messages: vec![(loc, format!("Expected type {expected}, but got {actual}"))],
            },
            InfWhereFinExpected => Report {
                registry,
                title: "Wrong boundedness".into(),
                messages: vec![(loc, "Expected finite type, but got infinite type".into())],
            },
            NameAlreadyDeclared { name } => Report {
                registry,
                title: "Name used twice".into(),
                messages: vec![(loc, format!("Name {name} was already declared"))],
            },
            CouldNotReadFile { error } => Report {
                registry,
                title: "Could not read file".into(),
                messages: vec![(loc, format!("<< {error} >>"))],
            },
        }
    }
}
// }}}

// display report {{{
impl Display for Report<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        let (Location(file, ref range), _) = self.messages[0];
        let path = self.registry.get_path(file);
        let bytes = self.registry.get_bytes(file);
        let line = 1 + bytes[range.start..].iter().filter(|c| b'\n' == **c).count();
        writeln!(f, "{}:{line}: {}", path.display(), self.title)?;

        for (_, msg) in &self.messages {
            writeln!(f, " -> {}", msg)?;
        }
        Ok(())
    }
}
// }}}
