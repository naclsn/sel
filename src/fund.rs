//! fundamental operations

use std::fmt::{Display, Formatter, Result as FmtResult};
use std::rc::Rc;

use crate::types::Type;

/// macro to generate the fn generating the type, internal to `mkmkty!`
macro_rules! mkmktyty {
    ($st:ident, Num) => {
        Type::number()
    };

    ($st:ident, Str+ $f:literal $(& $both:literal)? $(| $either:literal)?) => {
        {
            let f = mkmktyty!($st, + $f $(& $both)? $(| $either)?);
            Type::bytes(f)
        }
    };
    ($st:ident, Str) => {
        mkmktyty!($st, Str+ 0)
    };

    ($st:ident, $n:ident) => {
        $st.$n.clone()
    };

    ($st:ident, [$($i:tt)+]+ $f:literal $(& $both:literal)? $(| $either:literal)?) => {
        {
            let f = mkmktyty!($st, + $f $(& $both)? $(| $either)?);
            let i = mkmktyty!($st, $($i)+);
            Type::list(f, i)
        }
    };
    ($st:ident, [$($i:tt)+]) => {
        mkmktyty!($st, [$($i)+]+ 0)
    };

    ($st:ident, (($($a:tt)+), $($b:tt)+)) => {
        {
            let f = mkmktyty!($st, ($($a)+));
            let s = mkmktyty!($st, $($b)+);
            Type::pair(f, s)
        }
    };
    ($st:ident, ($a:tt, $($b:tt)+)) => {
        mkmktyty!($st, (($a), $($b)+))
    };
    ($st:ident, ($a:tt+ $f:literal $(& $both:literal)? $(| $either:literal)?, $($b:tt)+)) => {
        mkmktyty!($st, (($a+ $f $(& $both)? $(| $either)?), $($b)+))
    };

    ($st:ident, ($($t:tt)+)) => {
        mkmktyty!($st, $($t)+)
    };

    ($st:ident, $p:tt -> $($r:tt)+) => {
        {
            let p = mkmktyty!($st, $p);
            let r = mkmktyty!($st, $($r)+);
            Type::func(p, r)
        }
    };
    ($st:ident, $p:tt+ $f:literal $(& $both:literal)? $(| $either:literal)? -> $($r:tt)+) => {
        mkmktyty!($st, ($p+ $f $(& $both)? $(| $either)?) -> $($r)+)
    };

    ($st:ident, + $n:literal) => {
        false //$st._fin[$n]
    };
    ($st:ident, + $l:literal&$r:literal) => {
        false //$st.types.both($st._fin[$l], $st._fin[$r])
    };
    ($st:ident, + $l:literal|$r:literal) => {
        false //$st.types.either($st._fin[$l], $st._fin[$r])
    };
}

/// macro to generate the fn generating the type: (nb_infinites, names; type)
macro_rules! mkmkty {
    ($f:tt$(, $n:ident)*; $($ty:tt)+) => {
        || {
            struct State {
                //_fin: [Boundedness; $f+1],
                $($n: Rc<Type>,)*
            }
            let _state = State {
                //_fin: mkmkty!(@ types $f),
                $($n: Type::named(stringify!($n).into()),)*
            };
            mkmktyty!(_state, $($ty)+)
        }
    };

    //(@ $types:ident 0) => { [$types.finite(true)] };
    //(@ $types:ident 1) => { [$types.finite(true), $types.finite(false)] };
    //(@ $types:ident 2) => { [$types.finite(true), $types.finite(false), $types.finite(false)] };
    //(@ $types:ident 3) => { [$types.finite(true), $types.finite(false), $types.finite(false), $types.finite(false)] };
}

macro_rules! make {
    ($($(#[$desc:meta])+ $var:ident($name:literal) :: $mkty:expr),*$(,)?) => {
        /// Fundamental operations that cannot be implemented in the language itself.
        ///
        /// This does not mean everything else can be done within the language;
        /// eg OS-level interractions are to be implemented somewhere else. Likewise,
        /// not all maths may be accessible with that limited of a set (idk really).
        ///
        /// It basically represent what an interpreter will in any way need to implement.
        /// Technically `pipe`, `tonum` and `tostr` _could_ be implemented in the prelude
        /// but are not for consistency with other syntax-level functions.
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub enum Fund { $($(#[$desc])+ $var,)* }

        impl Display for Fund {
            fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
                write!(f, "{}", match self { $(Self::$var => $name,)* })
            }
        }

        impl Fund {
            pub fn try_from_name(name: &str) -> Option<Self> {
                match name { $($name => Some(Self::$var),)* _ => None }
            }

            pub fn make_type(&self) -> Rc<Type> {
                match self { $(Self::$var => ($mkty as fn() -> Rc<Type>)(),)* }
            }

            pub fn desc(&self) -> &'static [&'static str] {
                match self { $(Self::$var => &[$(stringify!($desc),)+],)+ }
            }
        }
    };
}

make! {
    /// the input
    Stream("-")                  :: mkmkty!(1          ; Str+1                   ),
    /// the type-restricting function
    Astypeof("astypeof")         :: mkmkty!(0, a       ; a -> a -> a             ),
    /// make a list with the head element first, then the rest as tail;
    /// 'snoc lst val' is equivalent to the syntax '{val,, lst}'
    // we use snoc and not cons as it makes better error messages (because it binds the list first)
    Snoc("snoc")                 :: mkmkty!(1, a       ; [a]+1 -> a -> [a]+1     ),
    /// panics (runtime abort with message)
    Panic("panic")               :: mkmkty!(0, a       ; Str -> a                ),
    /// pipe two functions;
    /// 'pipe one two' is equivalent to the syntax 'one, two' ie 'two(one(..))'
    /// (see also 'compose')
    Pipe("pipe")                 :: mkmkty!(0, a, b, c ; (a->b) -> (b->c) -> a->c),
    /// convert a string into number;
    /// accept an infinite string for convenience but stop on the first invalid byte
    Tonum("tonum")               :: mkmkty!(1          ; Str+1 -> Num            ),
    /// convert a number into string
    Tostr("tostr")               :: mkmkty!(0          ; Num -> Str              ),

    /// make a list of numbers with the 8 bits bytes
    Bytes("bytes")               :: mkmkty!(1          ; Str+1 -> [Num]+1        ),
    /// make a list of numbers with the 32 bits codepoints
    Codepoints("codepoints")     :: mkmkty!(1          ; Str+1 -> [Num]+1        ),
    /// make a list of strings with the potentially multi-codepoints graphemes
    Graphemes("graphemes")       :: mkmkty!(1          ; Str+1 -> [Str]+1        ),
    /// make a string from the 8 bits bytes
    Unbytes("unbytes")           :: mkmkty!(1          ; [Num]+1 -> Str+1        ),
    /// make a string from the 32 bits codepoints
    Uncodepoints("uncodepoints") :: mkmkty!(1          ; [Num]+1 -> Str+1        ),
    /// make a string from the potentially multi-codepoints graphemes
    Ungraphemes("ungraphemes")   :: mkmkty!(1          ; [Str]+1 -> Str+1        ),

    /// add two numbers
    Add("add")                   :: mkmkty!(0          ; Num -> Num -> Num       ),
    /// invert a number
    Invert("invert")             :: mkmkty!(0          ; Num -> Num              ),
    /// multiply two numbers
    Mul("mul")                   :: mkmkty!(0          ; Num -> Num -> Num       ),
    /// negate a number
    Negate("negate")             :: mkmkty!(0          ; Num -> Num              ),
    /// sign of a number, in {-1, 0, 1}
    Signum("signum")             :: mkmkty!(0          ; Num -> Num              ),
    /// truncation (rounding toward zero)
    Trunc("trunc")               :: mkmkty!(0          ; Num -> Num              ),

    /// the inverse sine function
    Asin("asin")                 :: mkmkty!(0          ; Num -> Num              ),
    /// the exponent function
    Exp("exp")                   :: mkmkty!(0          ; Num -> Num              ),
    /// the natural logarithm function
    Log("log")                   :: mkmkty!(0          ; Num -> Num              ),
    /// the sine function
    Sin("sin")                   :: mkmkty!(0          ; Num -> Num              ),
}
