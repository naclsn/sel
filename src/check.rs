//! checking and lowering into an AST

use std::collections::HashMap;

use crate::error::{self, Error};
use crate::fund::Fund;
use crate::parse::{Apply, ApplyBase, Pattern, Script, Value};
use crate::scope::{Entry, Location, Scoping, SourceRef};
use crate::types::{Type, TypeList, TypeRef};

#[derive(Debug, Clone)]
pub struct Tree {
    pub ty: TypeRef,
    pub val: TreeVal,
}

#[derive(Debug, Clone)]
pub enum TreeVal {
    Number(f64),
    Bytes(Box<[u8]>),
    Word(String, Refers),
    List, // represents an empty list, as a cons-sequence starter (ie nil)
    Pair(Box<Tree>, Box<Tree>),
    Apply(Box<Tree>, Vec<Tree>), // base can only be 'Word', 'Binding' or 'Fundamental', or the error list isn't empty
    Binding(Pattern, Box<Tree>, Box<Tree>), // garbage if `pat.is_irrefutable()`
    Fundamental(Fund),
}

#[derive(Debug, Clone)]
pub enum Refers {
    Binding(Location), // from an enclosing let binding
    File(SourceRef),   // from external (user) file
    Builtin(String),   // eg from prelude
    Fundamental,       // eg cons, panic, bytes...
    Missing,
}

pub struct Checker<'t, 's> {
    types: &'t mut TypeList,
    scope: &'s mut Scoping,
    errors: Vec<Error>,
}

impl<'t, 's> Checker<'t, 's> {
    pub fn new(types: &'t mut TypeList, scope: &'s mut Scoping) -> Self {
        Self {
            types,
            scope,
            errors: Vec::new(),
        }
    }

    pub fn errors(&self) -> &[Error] {
        &self.errors
    }

    fn apply(&mut self, func: Tree, arg: Tree) -> Tree {
        let ty = Type::applied(func.ty, arg.ty, &mut self.types).unwrap_or_else(|err| {
            self.errors.push(Error(
                todo!("loc for {err:?} (from apply not applying)"),
                err,
            ));
            match self.types.get(func.ty) {
                Type::Func(_, ret) => *ret,
                _ => self.types.named(format!("ret")),
            }
        });

        let val = match func.val {
            TreeVal::Apply(base, mut args) => {
                args.push(arg);
                TreeVal::Apply(base, args)
            }
            TreeVal::Word(_, _) | TreeVal::Binding(_, _, _) | TreeVal::Fundamental(_) => {
                TreeVal::Apply(Box::new(func), vec![arg])
            }
            TreeVal::Number(_) | TreeVal::Bytes(_) | TreeVal::List | TreeVal::Pair(_, _) => {
                TreeVal::Apply(Box::new(func), vec![arg])
            }
        };

        Tree { ty, val }
    }

    fn fundamental(&mut self, op: Fund) -> Tree {
        Tree {
            ty: op.make_type(self.types),
            val: TreeVal::Fundamental(op),
        }
    }

    pub fn check_script(&mut self, script: &Script) -> Tree {
        let val = self.check_apply(&script.head);

        if matches!(self.types.get(val.ty), Type::Func(_, _)) {
            // [f, g, h] -> [pipe f g, h] -> pipe [pipe f g] h
            script.tail.iter().fold(val, |acc, cur| {
                let then = self.check_apply(cur);
                let pipe = self.fundamental(Fund::Pipe);
                let part = self.apply(pipe, acc);
                self.apply(part, then)
            })
        } else {
            // [x, f, g, h] -> h [g [f x]]
            script.tail.iter().fold(val, |acc, cur| {
                let func = self.check_apply(cur);
                self.apply(func, acc)
            })
        }
    }

    fn check_pattern_type(
        &mut self,
        pat: &Pattern,
        names: &mut HashMap<String, (Location, TypeRef)>,
    ) -> TypeRef {
        match pat {
            Pattern::Number(_) => self.types.number(),

            Pattern::Bytes(_) => {
                let fin = self.types.finite(true);
                self.types.bytes(fin)
            }

            Pattern::Word(w) => {
                let ty = self.types.named(w.clone());
                if let Some((loc, ty)) = names.get(w) {
                    self.errors.push(todo!(
                        "already declared in pattern at {loc:?} with :: &{ty}"
                    ));
                    *ty
                } else {
                    names.insert(w.clone(), (todo!("loc"), ty));
                    ty
                }
            }

            Pattern::List(items, rest) => {
                let items: Box<_> = items
                    .iter()
                    .map(|it| self.check_pattern_type(it, names))
                    .collect();

                todo!("ty of Pattern::List\nitem types: {items:?}");
                // i kept the commented code below that refers to check_list_content;
                // i just rewrote Value::List handling in check_value in a way
                // that's more regular: it's always checked as if {...,, {}} ie it
                // always relies on `cons` which will dictate the behavior uniformly
                // however i'm left with that one now which idk how to then

                let ty = 0;
                //let has = self.check_list_content(items);
                //let fin = self.types.finite(rest.is_none());
                //let ty = self.types.list(fin, has);

                if let Some(w) = rest {
                    if let Some((loc, ty)) = names.get(w) {
                        self.errors.push(todo!(
                            "already declared in pattern at {loc:?} with :: &{ty}"
                        ));
                    } else {
                        names.insert(w.clone(), (todo!("loc"), ty));
                    }
                }
                ty
            }

            Pattern::Pair(fst, snd) => {
                let fst = self.check_pattern_type(fst, names);
                let snd = self.check_pattern_type(snd, names);
                self.types.pair(fst, snd)
            }
        }
    }

    fn check_binding_br(
        &mut self,
        pat: &Pattern,
        res: &Value,
        alt: &Value,
    ) -> (TypeRef, Tree, Tree) {
        let mut names = HashMap::new();
        let param = self.check_pattern_type(pat, &mut names);
        self.scope.push(names);

        let res = self.check_value(res);

        let alt = if pat.is_irrefutable() {
            Tree {
                ty: self.types.number(),
                val: TreeVal::Number(0.0),
            }
        } else {
            let alt = self.check_value(alt);
            if let Err(err) = Type::harmonize(res.ty, alt.ty, self.types) {
                self.errors.push(Error(
                    todo!("loc for {err:?} (from harmonize res&alt)"),
                    err,
                    //ErrorKind::ContextCaused {
                    //    error: Box::new(Error(result.loc.clone(), e)),
                    //    because: ErrorContext::LetFallbackTypeMismatch {
                    //        result_type: snapshot.frozen(res_ty),
                    //        fallback_type: snapshot.frozen(fallback.ty),
                    //    },
                    //},
                ));
            }
            alt
        };

        self.scope.pop();
        (self.types.func(param, res.ty), res, alt)
    }

    pub fn check_apply(&mut self, apply: &Apply) -> Tree {
        let base = match &apply.base {
            ApplyBase::Binding(pat, res, alt) => {
                let (ty, res, alt) = self.check_binding_br(pat, &res, &alt);
                Tree {
                    ty,
                    val: TreeVal::Binding(pat.clone(), Box::new(res), Box::new(alt)),
                }
            }
            ApplyBase::Value(val) => self.check_value(val),
        };

        apply.args.iter().fold(base, |acc, cur| {
            let arg = self.check_value(cur);
            self.apply(acc, arg)
        })
    }

    // not used, but kept for now because of the notes;
    // see in check_pattern_type's Pattern::List branch too
    fn check_list_content(&mut self, items: impl IntoIterator<Item = TypeRef>) -> TypeRef {
        let ty = self.types.named("item".into());
        for it in items {
            // so, unless I'm off, I think the way I did rest part for literal lists
            // will not allow eg {inf, fin,, {}} because it would go:
            //      cons fin {} :: fin
            //      cons inf fin :: type error!
            // looking again, I think that was already the case in previous implementation,
            // so now I'm wondering if I should also disallow when no rest for both
            // consistency and removing hunks from `types` (and almost idealy whole `types.rs`)
            //
            // but does that makes sens? isn't that annoying that it won't be possible to eg
            //      {:hello:, -}, unwords
            // but say in this specific case, maybe the miss is to say literal bytes is inf..?
            if let Err(err) = Type::harmonize(ty, it, &mut self.types) {
                self.errors
                    .push(todo!("error::list_type_mismatch(..) for {err:?}"));
            }
            // constructing always with `cons` seems like the least sane but most regular approach
            // (insane because that' a de-optimisation which will /have/ to be corrected further down the line)
            //
            // cons <fin> [<fin>] :: [<fin>] -- free
            // cons <inf> [<inf>] :: [<inf>] -- free
            // cons <fin> [<inf>] :: [<inf>] -- can never work through type sys only
            // cons <inf> [<fin>] :: [<inf>] -- works if (no-op) coersion [<fin>] -> [<inf>]
        }
        ty
    }

    pub fn check_value(&mut self, value: &Value) -> Tree {
        match value {
            Value::Number(n) => Tree {
                ty: self.types.number(),
                val: TreeVal::Number(*n),
            },

            Value::Bytes(b) => Tree {
                ty: {
                    let fin = self.types.finite(true);
                    self.types.bytes(fin)
                },
                val: TreeVal::Bytes(b.clone()),
            },

            Value::Word(w) => {
                let (ty, prov) = match self.scope.lookup(w) {
                    Some(Entry::Binding(binding)) => {
                        (binding.ty, Refers::Binding(binding.loc.clone()))
                    }
                    Some(Entry::Defined(defined)) => (
                        self.types.duplicate(defined.ty, &mut Default::default()),
                        Refers::File(defined.loc.0),
                    ),
                    Some(Entry::Fundamental(fund)) => {
                        (fund.make_type(self.types), Refers::Fundamental)
                    }
                    None => {
                        let ty = self.types.named(w.clone());
                        self.scope.missing(w.clone(), ty);
                        (ty, Refers::Missing)
                    }
                };
                Tree {
                    ty,
                    val: TreeVal::Word(w.clone(), prov),
                }
            }

            Value::Subscr(script) => self.check_script(script),

            Value::List(items, rest) => {
                let items: Vec<_> = items.iter().map(|it| self.check_apply(it)).collect();

                let rest = if let Some(rest) = rest {
                    self.check_apply(rest)
                } else {
                    // empty list, as if {...,, {}}
                    let has = self.types.named("item".into());
                    let fin = self.types.finite(true);
                    Tree {
                        ty: self.types.list(fin, has),
                        val: TreeVal::List,
                    }
                };

                // cons(items[0], .... cons(items[n-1], cons(items[n], rest)) .... )
                // note: reverse error order -- XXX: may not want that
                items.into_iter().rfold(rest, |acc, cur| {
                    let cons = self.fundamental(Fund::Cons);
                    let part = self.apply(cons, cur);
                    self.apply(part, acc)
                })
            }

            Value::Pair(fst, snd) => {
                let fst = Box::new(self.check_value(fst));
                let snd = Box::new(self.check_value(snd));
                Tree {
                    ty: self.types.pair(fst.ty, snd.ty),
                    val: TreeVal::Pair(fst, snd),
                }
            }
        }
    }
}
