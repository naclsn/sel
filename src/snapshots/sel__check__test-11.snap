---
source: src/check.rs
expression: "t(b\"add 1 :dda:\")"
---
(
    Tree {
        ty: Number,
        val: Apply(
            Tree {
                ty: Func(
                    Number,
                    Func(
                        Number,
                        Number,
                    ),
                ),
                val: Word(
                    "add",
                    Fundamental(
                        Add,
                    ),
                ),
            },
            [
                Tree {
                    ty: Number,
                    val: Number(
                        1.0,
                    ),
                },
                Tree {
                    ty: Bytes(
                        true,
                    ),
                    val: Bytes(
                        [
                            100,
                            100,
                            97,
                        ],
                    ),
                },
            ],
        ),
    },
    [
        Error(
            Location(
                "<test>",
                0..3,
            ),
            ContextCaused {
                error: Error(
                    Location(
                        "<test>",
                        6..11,
                    ),
                    ExpectedButGot {
                        expected: Number,
                        actual: Bytes(
                            true,
                        ),
                        as_of: [
                            Wanted(
                                Number,
                            ),
                        ],
                    },
                ),
                because: AsNthArgToNowTyped {
                    nth_arg: 2,
                    type_with_curr_args: Func(
                        Number,
                        Func(
                            Number,
                            Number,
                        ),
                    ),
                },
            },
        ),
    ],
)
